electrum_nmc/electrum/names.py

```diff
@@ -131,6 +133,127 @@ def validate_value_length(value: bytes):
+
+        # Decode the service ID from base32 into the decoded_service_id bytearray
+        decoded_service_id = base64.b32decode(address.encode(), decoded_service_id)
+
+        # Check decoded service ID has the expected length
+        if len(decoded_service_id) != V3_ONION_SERVICE_ID_RAW_SIZE:
+            raise ValueError("Invalid service ID length")
+
+        # Check the version byte
+        version_byte = decoded_service_id[V3_ONION_SERVICE_ID_VERSION_OFFSET]
+        if version_byte > 0x03:
+            raise ValueError(f"Warning: this is a v{version_byte} onion service, which is newer than the v3 onion services that Electrum-NMC knows how to validate. Please verify that this is what you really meant to enter.")
+
+        if version_byte != 0x03:
+            raise ValueError(f"Obsolete v{version_byte} onion service")
+            
+        # Extract the public key from the decoded service ID
+        public_key = bytearray(decoded_service_id[:ED25519_PUBLIC_KEY_SIZE])
+
+        # Calculate the truncated checksum
+        truncated_checksum = calc_truncated_checksum(public_key)
+
+        # Check if the truncated checksum matches the corresponding bytes in the decoded service ID
+        if truncated_checksum[0] != decoded_service_id[V3_ONION_SERVICE_ID_CHECKSUM_OFFSET] or \
+                truncated_checksum[1] != decoded_service_id[V3_ONION_SERVICE_ID_CHECKSUM_OFFSET + 1]:
+            raise ValueError("Invalid checksum")
+    
+    except Exception as e:
+        raise ValueError(f"Invalid onion address: {e}")
+
+def calc_truncated_checksum(public_key):
+    # Define the size of the hash in bytes
+    SHA256_BYTES = 256 // 8
+    hash_bytes = bytearray(SHA256_BYTES)
+
+    hasher = hashlib.sha3_256()
+    assert SHA256_BYTES == hasher.digest_size
+
+    # Calculate the checksum
+    hasher.update(b".onion checksum")
+    hasher.update(public_key)
+    hasher.update(bytes([0x03]))
+    hash_bytes = bytearray(hasher.digest())
+
+    return hash_bytes[:2]
+
+def validate_zeronet_address(address: str) -> None: # Validate P2PKH Address (Zero Net)
+    try:
+        addrtype, _ = b58_address_to_hash160(address)
+    except Exception as e:
+        raise ValueError(f"Invalid zeronet address: {e}")
+    
+    if addrtype != 0:
+        raise ValueError("Invalid address type.")
+
+def validate_ip_address(address: str) -> None:
+    try:
+        ipaddress.ip_address(address)
+    except Exception as e:
+        raise ValueError(f"Invalid IP address: {e}")
+
+def validate_i2p_address(address: str) -> None: # Naming rules were implemeted from https://geti2p.net/en/docs/naming
+    try:
+        # Check length
+        if len(address) > 67:
+            raise ValueError("Address length exceeds 67 characters")
+
+        # Check if it ends with '.b32.i2p'
+        if address.endswith('.b32.i2p'):
+            validate_base32(address)
+            return  # End function if base32 validation passes
```

Comment:
The base32 form is the only I2P form that we should accept. The other form being checked for here is the petname form, which has never been used with Namecoin and AFAICT shouldn't be.